blake2/src/impl_hacl/blake2s.rs

```diff
@@ -0,0 +1,471 @@
+
+const PARAM_LEN: usize = 8;
+const MAX_LEN: usize = 32;
+
+/// A builder for [`Blake2s`]. `T` determines whether
+pub struct Blake2sBuilder<'a, T> {
+    key: T,
+    personal: &'a [u8; PARAM_LEN],
+    salt: &'a [u8; PARAM_LEN],
+}
+
+impl<'a> Blake2sBuilder<'a, &'a ()> {
+    /// Creates the builder for an unkeyed hasher.
+    pub fn new_unkeyed() -> Self {
+        Self {
+            key: &(),
+            personal: &[0; PARAM_LEN],
+            salt: &[0; PARAM_LEN],
+        }
+    }
+
+    /// Constructs the [`Blake2s`] hasher for unkeyed hashes and dynamic digest length.
+    pub fn build_var_digest_len(
+        self,
+        digest_length: u8,
+    ) -> Result<Blake2s<ConstsStaticKey<0>>, Error> {
+        if digest_length < 1 || digest_length as usize > MAX_LEN {
+            return Err(Error::InvalidDigestLength);
+        }
+
+        let key_length = 0;
+
+        let kk = index {
+            key_length,
+            digest_length,
+            last_node: false,
+        };
+
+        let params = blake2_params {
+            digest_length,
+            key_length,
+            fanout: 1,
+            depth: 1,
+            leaf_length: 0,
+            node_offset: 0,
+            node_depth: 0,
+            inner_length: 0,
+            salt: self.salt,
+            personal: self.personal,
+        };
+
+        let key = params_and_key {
+            fst: &[params],
+            snd: &[],
+        };
+
+        Ok(Blake2s {
+            state: malloc_raw(kk, key),
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Constructs the [`Blake2s`] hasher for unkeyed hashes and static digest length.
+    pub fn build_const_digest_len<const OUT_LEN: usize>(
+        self,
+    ) -> Result<Blake2s<ConstsStaticKeyStaticOut<0, OUT_LEN>>, Error> {
+        if OUT_LEN < 1 || OUT_LEN > MAX_LEN {
+            return Err(Error::InvalidDigestLength);
+        }
+
+        let digest_length = OUT_LEN as u8;
+        let key_length = 0;
+
+        let kk = index {
+            key_length,
+            digest_length,
+            last_node: false,
+        };
+
+        let params = blake2_params {
+            digest_length,
+            key_length,
+            fanout: 1,
+            depth: 1,
+            leaf_length: 0,
+            node_offset: 0,
+            node_depth: 0,
+            inner_length: 0,
+            salt: self.salt,
+            personal: self.personal,
+        };
+
+        let key = params_and_key {
+            fst: &[params],
+            snd: &[],
+        };
+
+        Ok(Blake2s {
+            state: malloc_raw(kk, key),
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'a, const KEY_LEN: usize> Blake2sBuilder<'a, &'a [u8; KEY_LEN]> {
+    /// Creates the builder for an keyed hasher for keys where the length is known at compile
+    /// time.
+    pub fn new_keyed_static(key: &'a [u8; KEY_LEN]) -> Result<Self, Error> {
```

Comment:
Do we have `_static`, `_const`, and `_var` now? Either I don't understand what they are used for, or we could unify this a little.